ynput/OpenPype
# Tools: Experimental tools

## Goal
Easily add experimental tools into any host without need of rebuilding or rewriting implementation plugins for hosts where it's complicated until the tool is ready for production.

## Description
We've decided to add a dialog widget which we can modify dynamically through python so host implementation needs only single button which will trigger this dialog and all dynamic logic happens on python side.
Tools should be defined in simplest possible way and artists should enable them in local settings before they can use it in host.

## Changes
- added `experimental_tools` folder into `openpype.tools`
    - folder contains a `dialog.py` which contains Qt dialog logic and  `tools_def.py` which holds definition of experimental tools
- definitions of experimental tool requires on 4 needed information
    - identifier which must be unique to save/load local settings and store it's widget
    - label which will be shown in dialog
    - callback which will be triggered button click in dialog
    - tooltip which should describe the tool and is shown on button hover
    - optionally may have host filters as some tools may not be available for all hosts
- added action for experimental tools dialog to Maya and Nuke

### How to test
1. Empty experimental tools
- open maya/nuke and open "Experimental tools..." through OpenPype menu
    - dialog saying that there are no experimental tools atm. shoul show

2. With experimental tools
- experiemental tools are empty right now but there is commented example tool in `openpype/tools/experimental_tools/tools_def.py` just uncomment lines in `ExperimentalTools` inside `__init__` method
- open maya/nuke and open "Experimental tools..." through OpenPype menu
    - dialog with single button should show but which should be disabled
- open local settings through tray go to Experimental tools secion and enable the tool
- go back to host and focus on the experimental tools dialog (it should refresh state of tools on the fly) and the button should be enabled
- if you'll click it should print `Triggered tool` into console

### Notes
- I've put tool definitions into tools but if you have better idea where to put them it's totally doable
- if experimental tool is disabled the button in dialog is disabled too but you can find out that you have to enable it only by hovering over and reading tooltip maybe we should group enabled/disabled tools into separated widgets and add message describing how to enable tools to disabled section or show the message all the time?
- any message modifications are enabled

## TODO
- add dialog callback into all hosts

Depends on https://github.com/pypeclub/OpenPype/pull/2166 - Fix of stylesheet colors
